module/zfs/zfs_vfsops.c

```diff
@@ -972,6 +972,152 @@ zfs_id_overquota(zfsvfs_t *zfsvfs, uint64_t usedobj, uint64_t id)
 	    zfs_id_overobjquota(zfsvfs, usedobj, id));
 }
 
+static zfsvfs_kstats_t empty_zfsvfs_kstats = {
+	{ "writes",	KSTAT_DATA_UINT64 },
+	{ "nwritten",	KSTAT_DATA_UINT64 },
+	{ "reads",	KSTAT_DATA_UINT64 },
+	{ "nread",	KSTAT_DATA_UINT64 },
+};
+
+static int
+zfsvfs_kstats_update(kstat_t *ksp, int rw)
+{
+	zfsvfs_t *zfsvfs = ksp->ks_private;
+	zfsvfs_kstats_t *zs = ksp->ks_data;
+
+	ASSERT3P(zfsvfs->z_iokstat->ks_data, ==, zs);
+
+	if (rw == KSTAT_WRITE)
+		return (EACCES);
+
+	zs->zk_writes.value.ui64 =
+	    aggsum_value(&zfsvfs->z_aggsum_stats.zas_writes);
+	zs->zk_nwritten.value.ui64 =
+	    aggsum_value(&zfsvfs->z_aggsum_stats.zas_nwritten);
+	zs->zk_reads.value.ui64 =
+	    aggsum_value(&zfsvfs->z_aggsum_stats.zas_reads);
+	zs->zk_nread.value.ui64 =
+	    aggsum_value(&zfsvfs->z_aggsum_stats.zas_nread);
+
+	return (0);
+}
+
+static void
+zfsvfs_kstats_create(zfsvfs_t *zfsvfs)
+{
+	ASSERT3P(zfsvfs->z_iokstat, ==, NULL);
+
+	/*
+	 * There should not be anything wrong with having kstats for
+	 * snapshots. Since we are not sure how useful they would be
+	 * though nor how much their memory overhead would matter in
+	 * a filesystem with many snapshots, we skip them for now.
+	 */
+	if (zfsvfs->z_issnap)
+		return;
+
+	/*
+	 * We want the code and the behavior of ZFS to be consistent and
+	 * uniform between different OSes that ZFS runs on. Thus for the
+	 * name of this kstat we are limited by KSTAT_STRLEN as definied
+	 * in illumos which is less than what we have in Linux
+	 * (31 vs 255 characters).
+	 *
+	 * Thus, to keep things the same we distinguish handles by their
+	 * objset IDs prepended by the pool's load_guid. Note that both
+	 * numbers are formatted in hex to save character space.
+	 * See breakdown in the next comment.
+	 *
+	 * We use the pool's load_guid because it is guaranteed to not
+	 * change as long as the machine is running (unlike the current
+	 * GUID from the pool's config which could change upona reguid).
+	 */
+	char kstat_name[KSTAT_STRLEN];
+	int n = snprintf(kstat_name, sizeof (kstat_name), "%llx-%llx",
+	    (unsigned long long)spa_load_guid(dmu_objset_spa(zfsvfs->z_os)),
+	    (unsigned long long)dmu_objset_id(zfsvfs->z_os));
+
+	/*
+	 * At the time of this writing, KSTAT_STRLEN is 31 (in illumos),
```

Comment:
KSTAT_STRLEN = 255 for Linux

Reply:
If you read the comment all the way to the end, I do mention that KSTAT_STRLEN in Linux is currently 255, but also wanted to give some background on why I used the load_guid + objset ID in hex (again, trying to keep things uniform between Linux, Illumos, and BSD - 255 vs 31 & 31)